nvme_tcp.go

```diff
@@ -0,0 +1,658 @@
+/*
+ *
+ * Copyright © 2020 Dell Inc. or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package gobrick
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"path"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/dell/gobrick/internal/logger"
+	intmultipath "github.com/dell/gobrick/internal/multipath"
+	intscsi "github.com/dell/gobrick/internal/scsi"
+	"github.com/dell/gobrick/internal/tracer"
+	wrp "github.com/dell/gobrick/internal/wrappers"
+	"github.com/dell/gobrick/pkg/multipath"
+	"github.com/dell/gobrick/pkg/scsi"
+	"github.com/dell/goiscsi"
+	"github.com/dell/gonvme"
+	"golang.org/x/sync/semaphore"
+	"golang.org/x/sync/singleflight"
+)
+
+/*
+const (
+	iSCSIFailedSessionMinimumLoginRetryInterval = time.Second * 300
+	iSCSIWaitDeviceTimeoutDefault               = time.Second * 30
+	iSCSIWaitDeviceRegisterTimeoutDefault       = time.Second * 10
+	iSCSIMaxParallelOperationsDefault           = 5
+)*/
+
+type NVMeTCPConnectorParams struct {
+	// nvmeLib command will run from this chroot
+	Chroot string
+
+	// timeouts
+	// how long to wait for iSCSI session to become active after login
+	WaitDeviceTimeout                      time.Duration
+	WaitDeviceRegisterTimeout              time.Duration
+	FailedSessionMinimumLoginRetryInterval time.Duration
+	MultipathFlushTimeout                  time.Duration
+	MultipathFlushRetryTimeout             time.Duration
+
+	MultipathFlushRetries int
+	MaxParallelOperations int
+}
+
+type DevicePathResult struct {
+	devicePaths []string
+	nguid       string
+}
+
+func NewNVMeTCPConnector(params NVMeTCPConnectorParams) *NVMeTCPConnector {
+	mp := multipath.NewMultipath(params.Chroot)
+	s := scsi.NewSCSI(params.Chroot)
+
+	conn := &NVMeTCPConnector{
+		multipath: mp,
+		scsi:      s,
```

Comment:
Do we need `iscsi` in the `nvme` connection?

Reply:
this is scsi information